homeassistant/components/recorder/purge.py

```diff
@@ -47,7 +47,7 @@ def purge_old_data(
         # Purge a max of MAX_ROWS_TO_PURGE, based on the oldest states or events record
         event_ids = _select_event_ids_to_purge(session, purge_before)
         state_ids, attributes_ids = _select_state_and_attributes_ids_to_purge(
-            session, purge_before, event_ids
```

Comment:
The event ids are indexed and we already constrain on time_fired which is always the same as last_updated so there is no need to examine every last_updated value here